gtk-3-24 rebase (code)

uajain opened the pull request:
**Note**: Please ignore the branch naming (i.e. uajain/gtk-3-24-without-endless-theme) . I wish there was a way in Github to change the base branch for the PR.


=== Rebase list ===
drop f06af175d6 Don't let offscreen widget do grabbing
drop a5dd9efaf0 Do not allow devices in an offscreen hierarchy to take grabs.
drop 1946976252 Consistently treat gdkenumtypes.[hc] as source files
drop 8873220d95 Disable VIQR IM for vi locale
drop 72ed63a00e Don't list images from unknown directories in icon cache
drop 40cf7c4a63 disable a11y test accessibility-dump
drop 76afd69415 Mark known failing tests as non-fatal
drop a2f732622e gdk/x11: Check if we have access to GL before using GLX API
---- _Above patches came from debian/patches dir of v`3.22`_ ---

--- _Below patches are **re-applied** from debian/patches dir of v`3.24.7`_ ---
39d3352346 Revert "Update some a11y test results"
4ded4342bd Don't test default-constructed GdkPixbuf properties
95b3c4407d Mark known failing tests as non-fatal
043fb16b63 Don't list images from unknown directories in icon cache
6d12ef07cb Do not use VIQR input method for vi locale by default
2500eaa5c8 Consistently treat gdkenumtypes.[hc] as source files
018f1ab540 Do not allow devices in an offscreen hierarchy to take grabs.
6e18e08783 Don't let offscreen widget do grabbing

=== Rest of downstream changes ===
pick b18a389088 _mips_treeview_row_separator_height
(**Drop** as per wjt comment)
pick 4e85dec762 Remove Computer from file chooser
temporary_drop 97865737da theme: add Endless-specific customizations
pick ccfd545509 Comment out VS 2011/2012 project build
pick a659d4669f gtkwindow: add an environment variable to disable backdrop styling
(**Drop** as per cosimoc's comment below.)
temporary_drop 63abbc5aa4 Use EndlessOS icon theme as first fallback after checking the current theme

pick 77c7931cda filechooserwidget: make sure it fits in screen
(**Drop** as per wjt comment)
pick 9469fa2e7d filechooserwidget: allow to hide the sidebar
(**Drop** as per wjt comment)
pick 995901bfd8 filechooserwidget: reduce location width-chars
(**Drop** as per wjt comment)
pick 2216d406ff filechooserwidget: use a scrolled window for filters
(**Drop** as per wjt comment)

pick 35891e6fad TMP: Remove broken reftests
(**Dropped**)
pick b34f3f9847 gdk/x11: Add conditional support for EGL-X11
pick c9b12a2bdf gdk/x11: Force GLES when using EGL X11
pick 8d16fadeaf gl: Add private glReadPixels() wrapper
pick 2f1e958fdc Use the generic names for glFramebuffer functions
pick eebd1ccbf6 glarea: Don't use GL_RGBA8 with OpengGL ES
pick a73ddba203 gdk_gl_context_download_texture() convert pixel data from rgba to argb.
temporary_drop f29cbba682 theme: Add endless titlebar themes to gtk-contained-dark
(**squash** with 97865737da "Add Endless customizations")
pick bc933941cc GdkX11GLContext: x11-eglx, implemented texture_from_surface()
pick 9e39288523 x11: Check if the window is composited for GL frame sync
pick ba975b7c33 Use non standard _GTK_WINDOW_UNREDIRECTED
temporary_drop a1270f0ca4 theme: Set dark background for Chrome and Chromium > 59 top bar
pick 65fcec37e0 x11: listen to property changes on the root window
pick 547a0c15c6 x11: update screen size when workarea changes
pick 79cddb2f40 x11: only notify a new workarea when it changed

(Chromium patches are squashed with theme: a1270f0ca4: Set dark background for Chrome and Chromium > 59 top bar)
temporary_drop a4909276dd Chromium header bar: match default gradient
temporary_drop f4b9c0d20e Chromium header bar: specify backdrop gradient
temporary_drop 73ea976ab7 Chromium header bar: fix button color for Chrome 62
temporary_drop 2a9d1858d9 Chromium header bar: match default hover/pressed buttons
temporary_drop 0b9b8be4b1 Chromium header bar: regenerate CSS files
temporary_drop a762d81aca Chromium header bar: fix user button color for Chrome 64
temporary_drop d3b1a7c2ba Chromium header bar: regenerate CSS files

pick fee58ffbf8 tests: Fix build with external file
pick 502ad61c3c display: Avoid unsetting the DESKTOP_STARTUP_ID variable too late
pick d64d9be465 display: Add new virtual gdk_display_get_startup_notification_id() method.
pick 6fba047f45 application: Use the new API to get the startup notification ID
pick 770ccd30fd application: Complete startup notification sequence for remote invocations
temporary_drop 367b35a0d1 theme: remove special theming for mutter-topbar-shadow
(**squash** with 97865737da "Add Endless customizations")
temporary_drop 5583c3b071 theme: fix text readability on unfocused Chrome/Chromium tabs
(**Squash** with chromium patch set)
 

Most of them applied easily with few minor rebase conflicts.
~~`temporary_drop` denotes that we decided to hold off as per [ this comment](https://phabricator.endlessm.com/T25366#701521).~~
I have applied endless theme changes on this branch **after** the upstream rebase. It looks good to me. 
Note that github re-organizes commits by dates, hence, it does not show the order which I see on cli. :(
My cli version `git log` for this branch : https://paste.gnome.org/pn9410sjp

https://phabricator.endlessm.com/T25366

uajain commented:
Seems to build but `dch` error.
This is because I haven't pushed tags yet.

uajain commented:
bot, retest this please

wjt commented:
This appears to be based on upstream commit d151223b1b which is many commits ahead of 3.24.7 and 2 commits behind upstream tag 3.24.8. I suggest that if we are going to pick a commit which is not an upstream release, we should base this on c19ae9635e which is just 2 non-merge commits ahead of the 3.24.8 tag, one of which is a commit from this branch. It's the last commit before the build system was switched to Meson.

> === Rest of downstream changes ===
> pick [b18a389](https://github.com/endlessm/gtk/commit/b18a389088fc3156e61761444103a7cc692dff9e) _mips_treeview_row_separator_height

This is not a downstream change. The hint is the `Gbp-Pq:` line in the message. I think this is an old Debian change that should be dropped.

> pick [4e85dec](https://github.com/endlessm/gtk/commit/4e85dec7625da2c3dbd553a6f345e85cb0cdc200) Remove Computer from file chooser

OK

> pick [ccfd545](https://github.com/endlessm/gtk/commit/ccfd5455090d0e4499f1c6ccd5dc11af267cd504) Comment out VS 2011/2012 project build

OK

> pick [a659d46](https://github.com/endlessm/gtk/commit/a659d4669f760c21b461f7940fcb9cf67f053340) gtkwindow: add an environment variable to disable backdrop styling

OK. I was curious where this was used; it seems it's only eos-app-store which is the thing that slides in from the LHS when you right-click the desktop and pick "Add Website".

> pick [77c7931](https://github.com/endlessm/gtk/commit/77c7931cdab9435fb02e0ae2dad97f3b4a2c3470) filechooserwidget: make sure it fits in screen
> pick [9469fa2](https://github.com/endlessm/gtk/commit/9469fa2e7d3659263c554ce8839ae7ea45baf4ac) filechooserwidget: allow to hide the sidebar
> pick [995901b](https://github.com/endlessm/gtk/commit/995901bfd8ce943326dd5d5b4015fd7cd0d3333a) filechooserwidget: reduce location width-chars
> pick [2216d40](https://github.com/endlessm/gtk/commit/2216d406ff4a10bdc0919471f893bc030270c3bf) filechooserwidget: use a scrolled window for filters

These are all enough but if we were to declace 800×600 the minimum resolution, could we drop these? I would be curious to know just how small the screen has to get for these to kick in.

> pick [35891e6](https://github.com/endlessm/gtk/commit/35891e6fad5e41c7eb9890fea7eca758b0581bdd) TMP: Remove broken reftests

T[e]mp[orary]? The message says:

> This does the same thing as the corresponding commit on the Debian branch

Is there a corresponding commit on the Debian branch these days? If we drop this patch, do the tests still pass?

> pick [b34f3f9](https://github.com/endlessm/gtk/commit/b34f3f98472c99a5060dcd47003b7bd5382b7d72) gdk/x11: Add conditional support for EGL-X11
> pick [c9b12a2](https://github.com/endlessm/gtk/commit/c9b12a2bdf8c8b705e28b45a28bc03e0f5ab9391) gdk/x11: Force GLES when using EGL X11
> pick [8d16fad](https://github.com/endlessm/gtk/commit/8d16fadeafdd17e48a93111bfc7f7c00c29d1c4d) gl: Add private glReadPixels() wrapper
> pick [2f1e958](https://github.com/endlessm/gtk/commit/2f1e958fdc46cd9554a9a2527bdce4978c9fac44) Use the generic names for glFramebuffer functions
> pick [eebd1cc](https://github.com/endlessm/gtk/commit/eebd1ccbf6b98de9c44a17dc5846ecfd4ab8cdc1) glarea: Don't use GL_RGBA8 with OpengGL ES
> pick [a73ddba](https://github.com/endlessm/gtk/commit/a73ddba2030a1e8754c02be1d685efbeeb6b9054) gdk_gl_context_download_texture() convert pixel data from rgba to argb.
> pick [bc93394](https://github.com/endlessm/gtk/commit/bc933941ccd951be9225304f01432d50f8dc5ef4) GdkX11GLContext: x11-eglx, implemented texture_from_surface()
> pick [9e39288](https://github.com/endlessm/gtk/commit/9e39288523498517fe12317b008c6ac8887bb900) x11: Check if the window is composited for GL frame sync

I don't have any useful input on these :-) I assume they applied cleanly!


> pick [ba975b7](https://github.com/endlessm/gtk/commit/ba975b7c335e94d9b5c290d4e74f100447bd0640) Use non standard _GTK_WINDOW_UNREDIRECTED

OK, I just reviewed the corresponding change in Mutter and now I know what it's for.


> pick [65fcec3](https://github.com/endlessm/gtk/commit/65fcec37e0b1898b43ead46ce00a547b6bd83905) x11: listen to property changes on the root window
> pick [547a0c1](https://github.com/endlessm/gtk/commit/547a0c15c6099ef73e1784ccc72faa995911bbd1) x11: update screen size when workarea changes
> pick [79cddb2](https://github.com/endlessm/gtk/commit/79cddb2f405aa6e9ef57bb27076f5d5e847c1e67) x11: only notify a new workarea when it changed

@cosimoc do you think these 3 are potentially upstreamable?


> pick [fee58ff](https://github.com/endlessm/gtk/commit/fee58ffbf88e51539eb29992d49e7751942d566e) tests: Fix build with external file

Looks potentially upstreamable, but OTOH since upstream has added Meson now, maybe there's not much point.

> pick [502ad61](https://github.com/endlessm/gtk/commit/502ad61c3c53d7fe88d30dd094980a9a72a0cafe) display: Avoid unsetting the DESKTOP_STARTUP_ID variable too late
> pick [d64d9be](https://github.com/endlessm/gtk/commit/d64d9be465a2f4ef054c6589e44d2e69ce8e883b) display: Add new virtual gdk_display_get_startup_notification_id() method.
> pick [6fba047](https://github.com/endlessm/gtk/commit/6fba047f454105c2d4f2bcb76fd102e510e0e7f8) application: Use the new API to get the startup notification ID
> pick [770ccd3](https://github.com/endlessm/gtk/commit/770ccd30fddbd2d695be5201848408f22a4454f7) application: Complete startup notification sequence for remote invocations

I see these were merged to GTK master for Gtk4. https://gitlab.gnome.org/GNOME/gtk/merge_requests/156. It may be worth suggesting backporting these to the stable branch.

----

(Thanks for grouping the theme changes together.)

> temporary_drop [9786573](https://github.com/endlessm/gtk/commit/97865737da254032f7ccaee6d944466a7be40ff8) theme: add Endless-specific customizations

OK.

> temporary_drop [63abbc5](https://github.com/endlessm/gtk/commit/63abbc5aa4951f7642376aa663de35bc5bea5a30) Use EndlessOS icon theme as first fallback after checking the current theme

OK.

> temporary_drop [f29cbba](https://github.com/endlessm/gtk/commit/f29cbba682e36fb07e2d4ed7d4bd04571074ba1b) theme: Add endless titlebar themes to gtk-contained-dark

I think this should be squashed into "add Endless-specific customizations"

> temporary_drop [367b35a](https://github.com/endlessm/gtk/commit/367b35a0d19ba201b6ad0680c53f71e8cc9cb50e) theme: remove special theming for mutter-topbar-shadow

Ditto.

> temporary_drop [a1270f0](https://github.com/endlessm/gtk/commit/a1270f0ca497e6af71e83974ae6ce396b38f4e8b) theme: Set dark background for Chrome and Chromium > 59 top bar
> temporary_drop [a490927](https://github.com/endlessm/gtk/commit/a4909276dd4c42d038935bba607bac7624e38ac4) Chromium header bar: match default gradient
> temporary_drop [f4b9c0d](https://github.com/endlessm/gtk/commit/f4b9c0d20e6281b7cf1e18430a43408a88074a03) Chromium header bar: specify backdrop gradient
> temporary_drop [73ea976](https://github.com/endlessm/gtk/commit/73ea976ab783ff4b8cc2861f235d449537bf258f) Chromium header bar: fix button color for Chrome 62
> temporary_drop [2a9d185](https://github.com/endlessm/gtk/commit/2a9d1858d95088ffe86c855998180f83344923fd) Chromium header bar: match default hover/pressed buttons
> temporary_drop [0b9b8be](https://github.com/endlessm/gtk/commit/0b9b8be4b1449bf988f443aa59518c9967346414) Chromium header bar: regenerate CSS files
> temporary_drop [a762d81](https://github.com/endlessm/gtk/commit/a762d81aca4d60d8b3b301347e3c257e51417c5f) Chromium header bar: fix user button color for Chrome 64
> temporary_drop [d3b1a7c](https://github.com/endlessm/gtk/commit/d3b1a7c2ba7ee3a7da7db9ad63a1a6f31b132f7d) Chromium header bar: regenerate CSS files
> temporary_drop [5583c3b](https://github.com/endlessm/gtk/commit/5583c3b071f7aca311071ee0a90670622686d505) theme: fix text readability on unfocused Chrome/Chromium tabs

These should be squashed together, IMO; or perhaps 1 commit that edits the .scss and another that generates the .css.

> (new commit) gtk/Makefile: Add generated gtktypefuncs.c to DISTCLEANFILES
> (Also tentatively accepted [upstream](https://gitlab.gnome.org/GNOME/gtk/merge_requests/732)).

OK!

cosimoc commented:
> > pick [65fcec3](https://github.com/endlessm/gtk/commit/65fcec37e0b1898b43ead46ce00a547b6bd83905) x11: listen to property changes on the root window
> > pick [547a0c1](https://github.com/endlessm/gtk/commit/547a0c15c6099ef73e1784ccc72faa995911bbd1) x11: update screen size when workarea changes
> > pick [79cddb2](https://github.com/endlessm/gtk/commit/79cddb2f405aa6e9ef57bb27076f5d5e847c1e67) x11: only notify a new workarea when it changed
> 
> @cosimoc do you think these 3 are potentially upstreamable?

I could have sworn I had submitted those upstream, but I can't find the bug now, so perhaps I didn't!
Yes, those are definitely meant to go upstream.

uajain commented:
>>temporary_drop a1270f0 theme: Set dark background for Chrome and Chromium > 59 top bar
temporary_drop a490927 Chromium header bar: match default gradient
temporary_drop f4b9c0d Chromium header bar: specify backdrop gradient
temporary_drop 73ea976 Chromium header bar: fix button color for Chrome 62
temporary_drop 2a9d185 Chromium header bar: match default hover/pressed buttons
temporary_drop 0b9b8be Chromium header bar: regenerate CSS files
temporary_drop a762d81 Chromium header bar: fix user button color for Chrome 64
temporary_drop d3b1a7c Chromium header bar: regenerate CSS files
temporary_drop 5583c3b theme: fix text readability on unfocused Chrome/Chromium tabs

>These should be squashed together, IMO; or perhaps 1 commit that edits the .scss and another that generates the .css.

Looking closely at commit messages of each of these, these changes spread across 4 phabricator tickets. I am not sure I feel comfortable squashing them into 1 or 2 commits. Maybe group them by tickets# ?

wjt commented:
> > > temporary_drop [a1270f0](https://github.com/endlessm/gtk/commit/a1270f0ca497e6af71e83974ae6ce396b38f4e8b) theme: Set dark background for Chrome and Chromium > 59 top bar
> > > temporary_drop [a490927](https://github.com/endlessm/gtk/commit/a4909276dd4c42d038935bba607bac7624e38ac4) Chromium header bar: match default gradient
> > > temporary_drop [f4b9c0d](https://github.com/endlessm/gtk/commit/f4b9c0d20e6281b7cf1e18430a43408a88074a03) Chromium header bar: specify backdrop gradient
> > > temporary_drop [73ea976](https://github.com/endlessm/gtk/commit/73ea976ab783ff4b8cc2861f235d449537bf258f) Chromium header bar: fix button color for Chrome 62
> > > temporary_drop [2a9d185](https://github.com/endlessm/gtk/commit/2a9d1858d95088ffe86c855998180f83344923fd) Chromium header bar: match default hover/pressed buttons
> > > temporary_drop [0b9b8be](https://github.com/endlessm/gtk/commit/0b9b8be4b1449bf988f443aa59518c9967346414) Chromium header bar: regenerate CSS files
> > > temporary_drop [a762d81](https://github.com/endlessm/gtk/commit/a762d81aca4d60d8b3b301347e3c257e51417c5f) Chromium header bar: fix user button color for Chrome 64
> > > temporary_drop [d3b1a7c](https://github.com/endlessm/gtk/commit/d3b1a7c2ba7ee3a7da7db9ad63a1a6f31b132f7d) Chromium header bar: regenerate CSS files
> > > temporary_drop [5583c3b](https://github.com/endlessm/gtk/commit/5583c3b071f7aca311071ee0a90670622686d505) theme: fix text readability on unfocused Chrome/Chromium tabs
> 
> > These should be squashed together, IMO; or perhaps 1 commit that edits the .scss and another that generates the .css.
> 
> Looking closely at commit messages of each of these, these changes spread across 4 phabricator tickets. I am not sure I feel comfortable squashing them into 1 or 2 commits. Maybe group them by tickets# ?

Personally I think the Chromium theming stuff should be one commit with links to all those Phabricator tickets. The reason there are many tickets, AFAICT, is that changes were needed over time as new Chrome versions were released. I don't think we need to keep that history around. But I don't feel that strongly about it.

One related thing: have you actually re-generated the CSS from the .scss source on this branch? It looks like if `sassc` is available at build time, it'll be run, but otherwise it's assumed that the developer ran `parse-sass.sh` and committed the result. I don't see `sassc` in the build-dependencies, and it looks like historically the generated code has been committed.

uajain commented:
>One related thing: have you actually re-generated the CSS from the .scss source on this branch? It looks like if sassc is available at build time, it'll be run, but otherwise it's assumed that the developer ran parse-sass.sh and committed the result. I don't see sassc in the build-dependencies, and it looks like historically the generated code has been committed.

No, I have not generated any CSS on my own. I am just working with whatever is there in the branch (i.e. generated CSS which is committed). 

I will squash the chromium headerbar changes to one commit and link all the phab tickets.

wjt commented:
> > One related thing: have you actually re-generated the CSS from the .scss source on this branch? It looks like if sassc is available at build time, it'll be run, but otherwise it's assumed that the developer ran parse-sass.sh and committed the result. I don't see sassc in the build-dependencies, and it looks like historically the generated code has been committed.
> 
> No, I have not generated any CSS on my own. I am just working with whatever is there in the branch (i.e. generated CSS which is committed).

Please regenerate the CSS.

uajain commented:
Addressed most of the comments:

Major Changes: 
* Rebase over GTK-3.24.8 and updated Debian packaging too (https://github.com/endlessm/gtk/pull/88)
* Dropped ` b18a389 _mips_treeview_row_separator_height`
* Dropped  `35891e6 TMP: Remove broken reftests` from this branch
* Squashed chromium header bars tickets to one commit -> regenerated CSS files and committed to the branch.

I am waiting to see if this branch builds successfully on jenkins with the corresponding packaging.

wjt commented:
rebuild this please

wjt commented:
retest this please!

cosimoc commented:
This patch was only used for the apps page in the old app center and can be removed: https://github.com/endlessm/gtk/commit/83da3ff0bd0e582f6a0e680a2f3a10f73eebee39

cosimoc commented:
Also not sure if we need to keep the file chooser patches now that we don’t support composite resolutions anymore.

uajain commented:
> This patch was only used for the apps page in the old app center and can be removed: [83da3ff](https://github.com/endlessm/gtk/commit/83da3ff0bd0e582f6a0e680a2f3a10f73eebee39)

Dropped.

>Also not sure if we need to keep the file chooser patches now that we don’t support composite resolutions anymore.

If "not sure", probably that is a  reason to preserve those commits?

cosimoc commented:
I mean, we are certainly dropping support for composite mode!

I was saying "not sure" because I haven't done all the scoping work to map what needs to be reverted when dropping composite mode! So I don't know if these patches were intentionally kept or forgotten by whoever did that scoping work :-)

wjt commented on gtk/theme/Adwaita/_endless.scss:
```diff
@@ -65,3 +65,55 @@ headerbar.default-decoration {
       }
   }
 }
+
```
Last URL in the commit message is missing the `h` of `https:`
